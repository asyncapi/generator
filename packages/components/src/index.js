export { Models } from './components/models';
