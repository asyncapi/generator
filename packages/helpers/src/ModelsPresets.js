const { JAVA_COMMON_PRESET } = require('@asyncapi/modelina');

// Helper function to add package declaration
function addPackage(content) {
  return `package com.asyncapi.models;\n\n${content}`;
}

const websocketJavaPreset = {
  class: {
    self({ content, model }) {
      // Solution to handle imports dynamically
      const requiredImports = new Set();
      requiredImports.add('import java.util.Objects;');
      Object.values(model.properties).forEach(property => {
        const type = property.property.type;

        if (type === 'Map<String, Object>') {
          requiredImports.add('import java.util.Map;');
        }
        // Add other type checks and imports as needed
      });
      const imports = Array.from(requiredImports).join('\n');
      return `package com.asyncapi.models;\n\n${imports}\n\n${content}`;
    }
  }
};

const otherModelTypes = ['enum', 'union'];
otherModelTypes.forEach(type => {
  websocketJavaPreset[type] = {
    self({ content }) {
      return addPackage(content);
    }
  };
});

const JavaModelsPresets = [
  {
    preset: JAVA_COMMON_PRESET,
    options: {
      equal: true,
      hashCode: true,
      classToString: true,
      marshalling: false
    }
  },
  websocketJavaPreset
];

module.exports = { JavaModelsPresets };