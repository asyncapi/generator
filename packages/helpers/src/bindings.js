/**
 * Extracts default query parameters from a channel’s WebSocket binding.
 *
 * @param {Map<string,string>} channels - A Map representing all AsyncAPI channels.
 * @returns {Map<string,string>} A Map whose keys are parameter names and whose values are their defaults (or `''`).
 *
 * @example
 * // Suppose channel.bindings().get("ws").values() returns:
 * // { query: { properties: { foo: { default: 'bar' }, baz: {} } } }
 * const params = getQueryParams(channel);
 * console.log(params.get('foo')); // → 'bar'
 * console.log(params.get('baz')); // → ''
 */
function getQueryParams(channels) {
  // current implementation assumes there is always one channel
  // at the moment only WebSocket binding support query params and the use case for WebSocket is that there is always one channel per AsyncAPI document
  const channel = !channels.isEmpty() && channels.all().entries().next().value[1];

  const queryMap = new Map();
  
  const hasWsBinding = channel?.bindings?.().has('ws');

  if (!hasWsBinding) {
    return null;
  }

  const wsBinding = channel.bindings().get('ws');

  const query = wsBinding.value()?.query;
  //we do not throw error, as user do not have to use query params, we just exit with null as it doesn't make sense to continue with query building
  if (!query) {
    return null;
  }
  
  // Drill into the JSON Schema properties
  const properties = query.properties;
  if (!properties || typeof properties !== 'object') {
    return null;
  }
  
  // Populate the map, preserving defaults
  for (const [key, schema] of Object.entries(properties)) {
    const value = schema.default ?? '';
    queryMap.set(key, String(value));
  }

  return queryMap;
}

module.exports = {
  getQueryParams
};