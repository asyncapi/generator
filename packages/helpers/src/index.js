const { getServerUrl }  = require('./servers');
const { getClientName } = require('./utils');

module.exports = {
  getServerUrl,
  getClientName
};