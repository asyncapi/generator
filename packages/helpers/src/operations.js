
/**
 * Get messages related to the provided operation.
 * @param {object} operation - The AsyncAPI operation object. 
 * @throws {Error} If opeartion object is not provided or is invalid.
 * 
 * @returns {null} if there are no messages
 * @returns messages resulting from operation
 */
function getOperationMessages(operation) {
  if (!operation) {
    throw new Error('Operation object must be provided.');
  }
  const messages = operation.messages();
  if (messages.isEmpty()) {
    return null;
  }
  return messages.all();
}
/**
   * Get examples related to the provided message.
   * @param {object} message 
   * @returns {Array} - An array of examples for the provided message.
   * @returns {null} if there are no examples
   * @throws {Error} If message object is not provided or is invalid.
   */
  
function getMessageExamples(message) {
  if (!message) {
    throw new Error('Message object must be provided.');
  }
  const examples = message.examples();
  if (examples.isEmpty()) {
    return null;
  }
  return examples;
}
  
module.exports = {
  getOperationMessages,
  getMessageExamples,
};