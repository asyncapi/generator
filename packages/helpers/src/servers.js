/**
   * Get server URL from AsyncAPI server object.
   *
   * @param {object} server - The AsyncAPI server object.
   * 
   * return {string} - The server URL.
   */
module.exports.getServerUrl = (server) => {
  let url = server.host();

  //might be that somebody by mistake duplicated protocol info inside the host field
  //we need to make sure host do not hold protocol info
  if (server.protocol() && !url.includes(server.protocol())) {
    url = `${server.protocol()}://${url}`;
  }
    
  if (server.hasPathname()) {
    url = `${url}${server.pathname()}`;
  }
    
  return url;
};

//TODO: this separate file for helpers for servers represents approach to keep all helpers in separate files related to extractions of data from specific high level AsyncAPI objects. Here we will have more helpers for example related to variables extraction from servers, security, etc.
    