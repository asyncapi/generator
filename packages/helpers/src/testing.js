const { readFile, rm, readdir} = require('fs/promises');
const path = require('path');

/**
 * Helper function to clean up test result paths recursively.
 * @param {Object} config - The configuration object containing paths to clean. 
 */
async function cleanTestResultPaths(config) {
  if (!config) return;

  if (config.testResultPath) {
    try {
      await rm(config.testResultPath, { recursive: true, force: true });
    } catch (error) {
      console.error(`Failed to clean ${config.testResultPath}: ${error.message}`);
    }
  } else if (!hasNestedConfig(config)) {
    console.warn('Configuration missing testResultPath - no test results to clean:', config);
  }

  // Recursively check nested configurations
  for (const subConfig of Object.values(config)) {
    if (typeof subConfig === 'object' && subConfig !== null) {
      await cleanTestResultPaths(subConfig);
    }
  }
}

// Helper function to check if config has nested sub-configs
function hasNestedConfig(obj) {
  return Object.values(obj).some(val => typeof val === 'object' && val !== null);
}

/**
 * Helper function to recursively get all files and directories in a given directory.
 * @param {string} dir - The directory path to scan.
 */
async function getDirElementsRecursive(dir) {
  const elements = [];
  
  const items = await readdir(dir, { withFileTypes: true });
  
  for (const item of items) {
    const fullPath = path.join(dir, item.name);
    if (item.isDirectory()) {
      // If it's a directory, add it to the list and recurse into it
      elements.push({
        type: 'directory',
        name: item.name,
        path: fullPath,
        children: await getDirElementsRecursive(fullPath)
      });
    } else {
      // If it's a file, just add it to the list
      elements.push({
        type: 'file',
        name: item.name,
        path: fullPath
      });
    }
  }

  return elements;
}

/**
 * Helper function to build parameters for the test cases.
 * 
 * Consider refactoring if language-specific logic grows.
 * 
 * @param {string} language - The target language (e.g., 'java', 'dart').
 * @param {Object} config - Configuration object containing test settings.
 * @param {Object} [baseParams={}] - Additional parameters to merge into the final set.
 * @returns {Object} - The final parameters object for use in the test case.
 */
function buildParams(language, config, baseParams = {}) {
  // Note: This function is currently hardcoded to treat 'java' differently by excluding the 'clientFileName' parameter.
  const isJava = language.toLowerCase() === 'java';

  return {
    server: 'echoServer',
    ...(isJava ? {} : { clientFileName: config.clientFileName }),
    ...baseParams,
  };
}

/**
 * Helper function to verify the directory structure against expected elements.
 * @param {Array} expectedElements - The expected elements in the directory.
 * @param {string} dirPath - The path of the directory to verify.
 */
async function verifyDirectoryStructure(expectedElements, dirPath) {
  for (const element of expectedElements) {
    const filePath = path.join(dirPath, element.name);

    if (element.type === 'directory') {
      // Recurse into the subdirectory
      await verifyDirectoryStructure(element.children, filePath);
    } else if (element.type === 'file') {
      // Verifying a file
      try {
        const content = await readFile(filePath, 'utf8');
        expect(content).toMatchSnapshot(element.name);
      } catch (err) {
        throw new Error(`File ${filePath} not found or couldn't be read. Original error: ${err.message}`);
      }
    }
  }
}

/*
 * Get the list of files in a directory
 *
 * @param {string} dir - The directory path.
 * 
 * return {Promise<string[]>} - A promise that resolves to an array of file paths.
 */
const listFiles = async (dir) => {
  const dirElements = await readdir(dir, { withFileTypes: true });

  // Filter to only files, map to full paths
  return dirElements
    .filter(dirE => dirE.isFile())
    .map(dirE => dirE.name);
};

module.exports = {
  cleanTestResultPaths,
  getDirElementsRecursive,
  verifyDirectoryStructure,
  buildParams,
  listFiles
};