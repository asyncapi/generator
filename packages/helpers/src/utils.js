/**
 * Get client name from AsyncAPI info.title
 *
 * @param {object} info - The AsyncAPI info object.
 * 
 * return {string} - The client name with "Client" appended at the end.
 */
module.exports.getClientName = (info) => {
  const title = info.title();
  
  // Remove spaces, make the first letter uppercase, and add "Client" at the end
  return `${title.replace(/\s+/g, '') // Remove all spaces
    .replace(/^./, char => char.toUpperCase()) // Make the first letter uppercase
  }Client`;
};
  