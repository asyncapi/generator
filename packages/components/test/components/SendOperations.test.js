import path from 'path';
import { render } from '@asyncapi/generator-react-sdk';
import { Parser, fromFile } from '@asyncapi/parser';
import { SendOperations } from '../../src/index';

const parser = new Parser();
const asyncapi_v3_path = path.resolve(__dirname, '../__fixtures__/asyncapi-v3.yml');

describe('Testing of SendOperation function', () => {
  let parsedAsyncAPIDocument;

  beforeAll(async () => {
    const parseResult = await fromFile(parser, asyncapi_v3_path).parse();
    parsedAsyncAPIDocument = parseResult.document;
  });

  test('render js websockets with send operations and client name', () => {
    const result = render(
      <SendOperations 
        language='javascript'
        clientName='AccountServiceAPI'
        sendOperations={parsedAsyncAPIDocument.operations().filterBySend()} 
      />
    );
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });

  test('render websockets without send operations', () => {
    const result = render(
      <SendOperations 
        language='python'
        clientName='AccountServiceAPI'
        sendOperations={parsedAsyncAPIDocument.operations().filterBySend()}
      />
    );
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });

  test('render js websockets with null send operations', () => {
    const result = render(
      <SendOperations 
        language='javascript'
        clientName='AccountServiceAPI'
        sendOperations={null} 
      />
    );
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });

  test('render js websockets without send operations', () => {
    const result = render(
      <SendOperations 
        language='javascript'
        clientName='AccountServiceAPI'
        sendOperations={[]} 
      />
    );
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });
});
