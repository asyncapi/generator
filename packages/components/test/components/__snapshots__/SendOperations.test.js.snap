// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing of SendOperation function render js websockets with null send operations 1`] = `""`;

exports[`Testing of SendOperation function render js websockets with send operations and client name 1`] = `
"/**
   * Sends a sendUserSignedup message over the WebSocket connection.
   * 
   * @param {Object} message - The message payload to send. Should match the schema defined in the AsyncAPI document.
   * @param {WebSocket} [socket] - The WebSocket connection to use. If not provided, the client's own connection will be used.
   * @throws {TypeError} If message cannot be stringified to JSON
   * @throws {Error} If WebSocket connection is not in OPEN state
   */
  static sendUserSignedup(message, socket) {
    try {
      socket.send(JSON.stringify(message));
    } catch (error) {
      console.error('Error sending sendUserSignedup message:', error);
    }
  }

  /**
   * Instance method version of sendUserSignedup that uses the client's own WebSocket connection.
   * @param {Object} message - The message payload to send
   * @throws {Error} If WebSocket connection is not established
   */
  sendUserSignedup(message){
    if(!this.websocket){
      throw new Error('WebSocket connection not established. Call connect() first.');
    }
    AccountServiceAPI.sendUserSignedup(message, this.websocket);
  }"
`;

exports[`Testing of SendOperation function render js websockets without send operations 1`] = `""`;

exports[`Testing of SendOperation function render websockets without send operations 1`] = `
"@staticmethod
  async def send_user_signedup_static(message, socket):
      \\"\\"\\"
      Send a send_user_signedup message using a provided WebSocket connection, without needing an instance.

      Args:
          message (dict or str): The message to send.
          socket (websockets.WebSocketCommonProtocol): The WebSocket to send through.

      Raises:
          Exception: If sending fails or the socket is not connected.
      \\"\\"\\"
      await AccountServiceAPI._send(message, socket)

  async def send_user_signedup(self, message):
      \\"\\"\\"
      Send a send_user_signedup message using the WebSocket connection attached to this instance.

      Args:
          message (dict or str): The message to send. Will be serialized to JSON if it's a dictionary.

      Raises:
          Exception: If sending fails or the socket is not connected.
      \\"\\"\\"
      await self._send(message, self.ws_app)"
`;
